static/src/javascripts/lib/cmp.spec.js

```diff
@@ -0,0 +1,215 @@
+// @flow
+import { init, onConsentNotification, consentState, _ } from 'lib/cmp';
+import { getAdConsentState as _getAdConsentState } from 'common/modules/commercial/ad-prefs.lib';
+
+const getAdConsentState: any = _getAdConsentState;
+
+jest.mock('common/modules/commercial/ad-prefs.lib', () => ({
+    getAdConsentState: jest.fn(),
+    thirdPartyTrackingAdConsent: jest.fn(),
+}));
+
+describe('cmp', () => {
```

Comment:
I'm not crazy about how these tests don't document the expected usage of the library. The library exposes callbacks for each purpose and upon invocation those callbacks are given the current state. The tests take a different approach. They register a jest function as the callback and then do a lookup of the whole consent to check the value was correct.

It'd be good to make sure the parameters passed to the callbacks are correct, but also don't underestimate the value of tests as documentation. In this case, documentation of how we expect developers to use the library. As written, I'd expect to see future code appear in frontend that registers handlers and then does a lookup of the whole consent, rather than reading the callback parameter.

Reply:
```js
const myCallBack = jest.fn();
onConsentNotification('advertisement', myCallBack);
```

The reason `myCallBack` is a jest mock function is so that I then can assert against it using jest to check things such as how many times it's been called and the value of the arguments it was called with. I could pass a callback function  instead which would look more like an actual implementation but I'd need to jump through more hoops to validate it was called as expected, which would involve a lot more code.

Reply:
If it's possible in the testing framework, we'd want to provide a callback where we can assert the following two properties:

* the callback was invoked
* the argument to that callback was correct

At the moment we're only verifying the former. As mentioned, a useful side-effect would be documenting the intended use.

If it's tricky then maybe we're stuck as is, but I do think the values passed to these callbacks are a very important part of the API (arguably the most fundamental piece). This means it's really important to add a test to make sure it stays correct.

Reply:
> If it's possible in the testing framework, we'd want to provide a callback where we can assert the following two properties:
> - the callback was invoked
> - the argument to that callback was correct

Unfortunately you can only run those expectation by using a jest mock function, however I do believe the expectations were asserting against cover both points you mentioned:

```js
expect(myCallBack).toHaveBeenCalledTimes(1); // test the callback was invoked
expect(myCallBack).toBeCalledWith(true); // test the argument to that callback was correct
```